File: 프로그래머스/Lv.1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.*;

class Solution {
    public int solution(String dartResult) {
        // 3번
        // 점수: 0 ~ 10
        // 보너스: S, D, T
        // 옵션: *, #

        List<Integer> nums = new ArrayList<>();

        for (int i = 0; i < dartResult.length(); i++) {
            if (Character.isDigit(dartResult.charAt(i))) {
                if (dartResult.substring(i, i + 2).equals("10")) {  // check 10
                    nums.add(10);
                    i++;
                } else {
                    nums.add(Integer.valueOf(dartResult.substring(i, i + 1)));
                }
            } else if ("SDT".contains(dartResult.substring(i, i + 1))) {
                calculate(dartResult.substring(i, i + 1), nums);
            } else {
                bonus(dartResult.substring(i, i + 1), nums);
            }
        }

        return nums.stream()
            .mapToInt(Integer::intValue)
            .sum();
    }

    static void calculate(String s, List<Integer> nums) {
        int idx = nums.size() - 1;
        int num = nums.get(idx);

        if (s.equals("S")) {
            nums.set(idx, (int) Math.pow(num, 1));
        } else if (s.equals("D")) {
            nums.set(idx, (int) Math.pow(num, 2));
        } else if (s.equals("T")) {
            nums.set(idx, (int) Math.pow(num, 3));
        }
    }

    static void bonus(String s, List<Integer> nums) {
        if (s.equals("*")) {
            if (nums.size() == 1) {
                nums.set(0, nums.get(0) * 2);
            } else {
                int idx1 = nums.size() - 2;
                int idx2 = nums.size() - 1;

                nums.set(idx1, nums.get(idx1) * 2);
                nums.set(idx2, nums.get(idx2) * 2);
            }
        } else if (s.equals("#")) {
            int idx = nums.size() - 1;
            nums.set(idx, nums.get(idx) * -1);
        }
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    l1 = ['0','1','2','3','4','5','6','7','8','9']
    l2 = ['S','D','T']
    l3 = ['*','#']
    
    lresult = []
    i = 0
    
    while i < len(dartResult):
        if dartResult[i] in l1: # 숫자인 경우
            if dartResult[i:i+2] == "10":
                lresult.append(10)
                i += 2
            else:
                lresult.append(int(dartResult[i]))
                i += 1
        elif dartResult[i] in l2:   # 점수인 경우
            if dartResult[i] == 'S':
                lresult[-1] = lresult[-1] ** 1
            elif dartResult[i] == 'D':
                lresult[-1] = lresult[-1] ** 2
            elif dartResult[i] == 'T':
                lresult[-1] = lresult[-1] ** 3
            i += 1
        elif dartResult[i] in l3:   # 보너스인 경우
            if dartResult[i] == '*':
                if len(lresult) > 1:
                    lresult[-1] = lresult[-1] * 2
                    lresult[-2] = lresult[-2] * 2
                else:
                    lresult[-1] = lresult[-1] * 2
            elif dartResult[i] == '#':
                lresult[-1] = lresult[-1] * -1
            i += 1
    
    return sum(lresult)